Add transmission-integration path and protocol

idelsink opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

This will add Protocol and Path support to the transmission integration. This will fix https://github.com/home-assistant/core/issues/101469.

See here an example of it running on my local machine:

![image](https://github.com/home-assistant/core/assets/8396902/64a18d34-7750-4f6d-a579-23f223c06af2)



## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes and closes #101469 
 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [x] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

idelsink committed:
Add transmission-integration path and protocol

This updates the config_flow migration from v1.1 to v1.2
including migration tests

engrbm87 reviewed (changes requested):
Few comments to consider:

- Changing config data schema requires updating existing entries. You need to add a method in `async_setup_entry` for the same.
- Consider using `CONF_SSL` checkbox to specify if `https` should be used.
- I don't think we can update `strings.json` file in this PR.

idelsink commented:
@engrbm87 Thanks for the feedback! :pray: 

I've changed it to the SSL setup as you suggested, that is indeed nicer. I was initially looking for something like this, but I probably missed that. I also removed the entry in the strings.json file, because that was no longer needed.

About the async_setup_entry, I'm currently looking into this in more detail how it works (still new to me). Am I supposed to create a `async_migrate_entry` method? Or something else?

engrbm87 commented on homeassistant/components/transmission/__init__.py:
```diff
@@ -210,12 +214,39 @@ async def async_unload_entry(hass: HomeAssistant, config_entry: ConfigEntry) ->
+        new = {**config_entry.data}
+
+        # add ssl
+        if CONF_SSL not in new:
```
If config version is 1 then definitely these keys won't be in `data`

idelsink replied:
So you would suggest by removing that check altogether  because it is practically meaningless?

engrbm87 commented on homeassistant/components/transmission/__init__.py:
```diff
@@ -210,12 +214,39 @@ async def async_unload_entry(hass: HomeAssistant, config_entry: ConfigEntry) ->
+        if CONF_SSL not in new:
+            new[CONF_SSL] = DEFAULT_SSL
+        # add path
+        if CONF_PATH not in new:
```
Same here

engrbm87 reviewed (changes requested):
Nice work. I think tests are needed as well.

idelsink commented:
> Nice work. I think tests are needed as well.

Thanks :pray:!

I see there is already a transmission test, let me see on how to integrate these migrations in there

idelsink commented:
@engrbm87 I've taken your feedback and:
- Removed the unnecessary checking
- Added a test to check if the migration went as expected; I'm not totally sure about the workflow for this, so I hope this is how it is supposed to work.
- I also moved the version to a constant so it can be used in the tests to check if the migrated config matches the required config

engrbm87 commented on tests/components/transmission/test_config_flow.py:
```diff
@@ -26,6 +32,18 @@ def mock_api():
         yield api
 
 
+async def test_config_flow_entry_migrate(hass: HomeAssistant) -> None:
```
Entry migration happens under `async_setup_entry` inside __init__.py. The tests should reflect the same.

engrbm87 commented on tests/components/transmission/test_init.py:
```diff
@@ -39,6 +46,18 @@ async def test_successful_config_entry(hass: HomeAssistant) -> None:
+    """Test that config flow entry is migrated correctly."""
+    entry = MockConfigEntry(domain=DOMAIN, data=MOCK_CONFIG_DATA_VERSION_1)
+
+    await transmission.async_migrate_entry(hass, entry)
```
You should not call the class method directly. Instead await `hass.config_entries.async_setup(entry.entry_id)` and ensure that the integration loaded successfully and that the config entry version is updated.

edenhaus replied:
Please implemented the above suggestion

idelsink replied:
I've changed it in the way you suggested! its loading the component and after that it checks the migration

edenhaus commented on homeassistant/components/transmission/config_flow.py:
```diff
@@ -37,7 +49,7 @@
     """Handle Tansmission config flow."""
 
-    VERSION = 1
+    VERSION = CONF_ENTRY_VERSION
```
Please increase only the minor version and leave the version as it is as this change is not a breaking change.
See https://developers.home-assistant.io/docs/config_entries_config_flow_handler#config-entry-migration

idelsink replied:
Ahh ok, I missed that there was a minor attribute in the version for components when I created it, I'll try to work on it soon when I have some time! Thanks for the feedback btw! :pray:

idelsink replied:
Updated accordingly!

edenhaus commented on homeassistant/components/transmission/strings.json:
```diff
@@ -4,10 +4,12 @@
           "username": "[%key:common::config_flow::data::username%]",
           "password": "[%key:common::config_flow::data::password%]",
-          "port": "[%key:common::config_flow::data::port%]"
+          "port": "[%key:common::config_flow::data::port%]",
```
Please add the attribute sorted correctly

idelsink replied:
Hey @edenhaus do you mean alpabatically sorted? Because if that's the case, I need to sort also entries that I did not add. 

If there is another sorting method being used, can you share me where to find this? Because the docs don't seem to mention one

emontnemery replied:
Yes, please sort existing items, existing and the ones you add, alphabetically 👍

idelsink replied:
Did this!

edenhaus commented on homeassistant/components/transmission/const.py:
```diff
@@ -23,14 +23,17 @@
     ),
 }
 CONF_ENTRY_ID = "entry_id"
+CONF_ENTRY_VERSION = 2
```
Please specify the version directly in config_flow.py

idelsink replied:
Changed it and its now in the config_flow

edenhaus commented:
This PR is a new feature and not a bugfix -> removed the bugfix label and checkmark

idelsink commented:
@edenhaus, @engrbm87; Besides the sorting issues where I left a comment, I've fixed the issues/suggestions as you both suggested! Thanks again! Let me know if there is anything that might need changing!

emontnemery commented on homeassistant/components/transmission/__init__.py:
```diff
@@ -210,12 +214,45 @@ async def async_unload_entry(hass: HomeAssistant, config_entry: ConfigEntry) ->
+            new[CONF_PATH] = DEFAULT_PATH
+            new[CONF_SSL] = DEFAULT_SSL
+
+        config_entry.version = 1
+        config_entry.minor_version = 2
+        hass.config_entries.async_update_entry(config_entry, data=new)
```
```suggestion
        hass.config_entries.async_update_entry(
            config_entry, data=new, version=1, minor_version=2
        )
```

idelsink replied:
Used your suggested change!

emontnemery commented on homeassistant/components/transmission/__init__.py:
```diff
@@ -210,12 +214,45 @@ async def async_unload_entry(hass: HomeAssistant, config_entry: ConfigEntry) ->
+    )
+
+    if config_entry.version == 1:
+        # Version 1.1 -> 1.2:
+        #   - Adding ssl
+        #   - Adding path
```
```suggestion
        # Version 1.2 adds ssl and path
```

idelsink commented:
All the issues should now be resolved!

emontnemery commented on homeassistant/components/transmission/strings.json:
```diff
@@ -5,9 +5,11 @@
-          "username": "[%key:common::config_flow::data::username%]",
           "password": "[%key:common::config_flow::data::password%]",
-          "port": "[%key:common::config_flow::data::port%]"
+          "path": "[%key:common::config_flow::data::path%]",
```
It's not clear what path this is.
Please add a `data_description` dictionary with a `path` key explaining to the user what this setting does.

idelsink replied:
Would it make sense to also add it to the rest, because if I add en explanation to one, it definitely looks weird (see the example) :')

![image](https://github.com/home-assistant/core/assets/8396902/0954f676-5866-401d-9de7-5af6c7f4ced4)

emontnemery replied:
Yes, of course it's beneficial if you add the description to all the fields 👍

emontnemery reviewed (changes requested):
Thanks @idelsink 👍 

Just one comment to improve the usability, then this PR is good to go.

emontnemery commented:
Please don't amend commits when you do changes. Instead, add additional commits to your PR branch and push. This project squashes PR commits when merging, so there will only be a single commit in the end.

emontnemery reviewed (approved):
Thanks, @idelsink 👍 

A follow-up PR adding data description to the other config flow fields would be welcome.
